token/program-2022/src/instruction.rs

```diff
@@ -1013,7 +1013,8 @@ pub fn initialize_account3(
     mint_pubkey: &Pubkey,
     owner_pubkey: &Pubkey,
 ) -> Result<Instruction, ProgramError> {
-    check_program_account(token_program_id)?;
+    // NOTE: There is no program account check here for associated-token-account
+    // compatibility with both token programs.
```

Comment:
This feels weird. Can we instead update `check_program_account()` to cover both program_ids?
Or should we use different instruction builders in ATA processor depending on the program_id?

Reply:
I did this to eventually support generic token programs, but we're likely a ways off from that. We can either use different instruction builders based on program id, or have token-2022 check for either `spl_token_2022::id()` or `spl_token::id()`.  Which do you prefer?

Reply:
I think I like the second one to avoid more branches and to be sure that token-2022 is always compatible with token

Reply:
I think I prefer the different instruction builders, if it's not too messy.

Reply:
> I think I like the second one to avoid more branches and to be sure that token-2022 is always compatible with token

Oh, hmm... let me ponder that another minute...

Reply:
You know, it makes me think... If that's the direction we're going, I wonder if we actually ought to pull spl-token into spl-token-2022 as a dependency, and not duplicate all the stuff we've c+p'ed.

Anyway, I guess I've talked myself down to no preference on the two approaches. I think the extended `check_program_account()` (which I think should be a separate method from the existing `check_program_account()`, incidentally) would require a dependency on spl_token regardless

Reply:
Actually, I'm now leaning strongly towards having spl-token-2022 doing the check for applicable instructions.  If we want other dapps and frontends to upgrade, they'll want backwards compatibility with all the instructions most likely, and not support just the new one. Otherwise they'll have to implement their own branching every time they create a new instruction.

Reply:
All right, that should be everything. We can reduce the copy-paste between spl-token and spl-token-2022 later. There isn't too much there surprisingly -- only `state.rs` really